packages/profile-sync-controller/src/controllers/user-storage/schema.ts

```diff
@@ -1,38 +1,74 @@
 import { createSHA256Hash } from './encryption';
 
-type UserStorageEntry = { path: string; entryName: string };
-
 /**
- * The User Storage Endpoint requires a path and an entry name.
- * Developers can provide additional paths by extending this variable below
+ * The User Storage Endpoint requires a feature name and a namespace key.
+ * Developers can provide additional features and keys by extending these types below.
  */
-export const USER_STORAGE_ENTRIES = {
-  notificationSettings: {
-    path: 'notifications',
-    entryName: 'notificationSettings',
-  },
-} satisfies Record<string, UserStorageEntry>;
 
-export type UserStorageEntryKeys = keyof typeof USER_STORAGE_ENTRIES;
+export const USER_STORAGE_SCHEMA = {
+  notifications: ['notificationSettings'],
+} as const;
+
+type UserStorageSchema = typeof USER_STORAGE_SCHEMA;
+type UserStorageFeatures = keyof UserStorageSchema;
+type UserStorageFeatureKeys<Feature extends UserStorageFeatures> =
+  UserStorageSchema[Feature][number];
+
+type UserStorageFeatureAndKey = {
+  feature: UserStorageFeatures;
+  key: UserStorageFeatureKeys<UserStorageFeatures>;
+};
+
+export type UserStoragePath = {
+  [K in keyof UserStorageSchema]: `${K}.${UserStorageSchema[K][number]}`;
+}[keyof UserStorageSchema];
+
+export const getFeatureAndKeyFromPath = (
+  path: UserStoragePath,
+): UserStorageFeatureAndKey => {
+  if (!/^\w+\.\w+$/u.test(path)) {
```

Comment:
regex scares me. Does the `w+` work for dashes/underscores? This might be "good enough", so can leave it.

Reply:
@Prithpal-Sooriya it works for underscores, but not for dashes. it kinda makes sense, but I'm ok with replacing w with `[a-zA-Z0-9]`. LMK what you feel would be better!